File: js/endless-binary/bin2dec-2.js
// taidalab Version 1.4.0
// https://github.com/taidalog/taidalab
// Copyright (c) 2022 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
function checkAnswerb2d2 (answer, question, last_answers) {
    // Getting the user input.
    const numberInput = document.getElementById('numberInput');
    const inputValue = escapeHtml(numberInput.value);
    console.log('inputValue : ' + inputValue);
    
    numberInput.focus();

    // Making an error message.
    const questionWithoutSpace = question.replace(' ', '');
    const errorMessage = newErrorMessageDec(questionWithoutSpace, inputValue);
    document.getElementById('errorArea').innerHTML = errorMessage;
    
    // Exits when the input was invalid.
    if (errorMessage) {
        return;
    }
    
    const inputValueAsInt = parseInt(inputValue);
    
    // Converting the input in order to use in the history message.
    const digit = 3;
    const spacePaddedInputValue = inputValue.padStart(digit, ' ').replace(' ', '&nbsp;');
    
    const sourceRadix = 2;
    const bin = inputValueAsInt.toString(sourceRadix);
    console.log('inputValue -> binary : ' + bin);

    // Making a new history and updating the history with the new one.
    const destinationRadix = 10;
    const outputArea = document.getElementById('outputArea');
    const currentHistoryMessage = newHistory((inputValueAsInt == answer), spacePaddedInputValue, destinationRadix, bin, sourceRadix);
    const historyMessage = concatinateStrings(currentHistoryMessage, outputArea.innerHTML);
    console.log(currentHistoryMessage);
    console.log(historyMessage);
    outputArea.innerHTML = historyMessage;
    
    if (inputValueAsInt == answer) {
        // Making the next question.
        let nextNumber = 0;
        
        console.log(last_answers);
        do {
            nextNumber = getRandomBetween(0, 255);
            console.log(nextNumber);
            console.log(last_answers.some((element) => element == nextNumber));
        } while (last_answers.some((element) => element == nextNumber));

        const nextBin = nextNumber.toString(sourceRadix);
        const splitBin = splitBinaryStringBy(4, nextBin);
        console.log(nextBin);
        console.log(splitBin);
        
        document.getElementById('questionSpan').innerText = splitBin;

        numberInput.value = '';

        // Updating `lastAnswers`.
        // These numbers will not be used for the next question.
        const answersToKeep = 10;
        const lastAnswers = [nextNumber].concat(last_answers).slice(0, answersToKeep);

        // Setting the next answer to the check button.
        document.getElementById('submitButton').onclick = function() { checkAnswerb2d2(nextNumber, splitBin, lastAnswers); return false; };
    }
}


function initBin2Dec2 () {
    // Initialization.
    const initNumber = getRandomBetween(0, 255);
    const initBin = initNumber.toString(2);
    const splitBin = splitBinaryStringBy(4,initBin);
    console.log(initNumber);
    console.log(initBin);
    console.log(splitBin);
    
    const sourceRadix = 2;
    const destinationRadix = 10;
    
    document.getElementById('questionSpan').innerText = splitBin;
    document.getElementById('binaryRadix').innerHTML = '<sub>(' + destinationRadix + ')</sub>';
    document.getElementById('srcRadix').innerHTML = '(' + sourceRadix + ')';
    document.getElementById('dstRadix').innerHTML = destinationRadix;
    document.getElementById('submitButton').onclick = function() { checkAnswerb2d2(initNumber, splitBin, [initNumber]); return false; };
}

File: src/endless-binary/Bin2Dec2.fs
// taidalab
// https://github.com/taidalog/taidalab
// Copyright (c) 2022-2025 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
namespace Taidalab.EndlessBinary

open Browser.Dom
open Taidalab
open Taidalab.Number
open Taidalab.EndlessBinary.Course
open Fermata.RadixConversion

module Bin2Dec2 =
    let help =
        """
        2進数から10進数への変換をエンドレスで練習できます。<br>
        出題範囲は n (0&le;n&le;255) です。<br>
        ヒントはありませんので、慣れてからどうぞ。
        """

    let question (lastAnswers: int list) : int =
        newNumber (fun _ -> getRandomBetween 0 255) (fun n -> List.contains n lastAnswers = false)

    let exec' (lastNumbers: int list) (answer: int) : unit =
        exec
            Dec.validate
            Bin2Dec1.error
            id
            Bin2Dec1.history'
            question
            (fun x -> splitBinaryStringBy 4 (Dec x |> Dec.toBin))
            id
            (fun _ -> "")
            (fun _ -> ())
            4
            lastNumbers
            answer

    let init () : unit =
        document.title <- "2進数→10進数 (2) - taidalab"

        let header = document.querySelector "header"
        header.innerHTML <- Content.Common.header
        header.className <- "bin2dec"

        (document.getElementById "hamburgerButton").onclick <-
            (fun _ ->
                (document.querySelector "nav").classList.toggle "flagged" |> ignore
                (document.getElementById "barrier").classList.toggle "flagged" |> ignore
                (document.querySelector "main").classList.toggle "flagged" |> ignore)

        (document.getElementById "barrier").onclick <-
            (fun _ ->
                (document.querySelector "nav").classList.remove "flagged" |> ignore
                (document.getElementById "barrier").classList.remove "flagged" |> ignore
                (document.querySelector "main").classList.remove "flagged" |> ignore)

        (document.querySelector "#headerTitle").innerHTML <-
            """<span>2進数→10進数 (2) - </span><span translate="no">taidalab</span>"""

        (document.querySelector "main").innerHTML <- EndlessBinary.Course.main help "help-color bin2dec"
        (document.querySelector "#submitButton").className <- "bin2dec"
        (document.querySelector "#questionArea").innerHTML <- Content.Common.question

        Bin2Dec1.init' question (fun _ -> "") (fun _ -> ()) exec' Course.keyboardshortcut
